### ISSUE TITLE
Extending call/fetch(Remote) functions

### ISSUE THREAD

**Issue by ghost** [multitheftauto/mtasa-blue#612]
**Is your feature request related to a problem? Please describe.**

At this moment, when you make a request to another host which takes years to response (ex: another MTA server which is running a mysql query using mysql module on a busy server) causes the calls to get stuck (timeout is disabled to wait for a response here) until they finally return the result. Sometimes, querys can get stuck and with it, the request / response will stuck the queue. I dont really think that you would generate randomly generated queueNames for remote calls.

**Describe the solution you'd like**

There should be some functions which can control the requests with the possibility of canceling them.

**Describe alternatives you've considered**

We could extend the current fetch/call(Remote) feature so that we can control it more:

1. call/fetchRemote functions should return an userdata with which you would be able to control your re
quests manually (when needed)
2. Add table getPendingNetworkCalls(string resource/userdata resource): With this function, we would be able to get detailed informations about each pending request(ex: url, resource, ...)
3. Add bool abortPendingRequest(userdata pendingRequest)
4. Add events for them: onNetworkRequestComplete(source = request, responseData, errorCode, abortedByScript = false)

**Additional context**

With this feature, we would've more abilities and possibilities to manage remote calls - which might also be good for debugging your server.

Also these functions / events might be good to have since debugging can sometimes be a pain.

I hope that you guys understand what i wanted to say with my request (which i would also work on once i got positive feedback) and i hope to get some feedback / ideas for this.

- Luxy.c

**Comment by ghost** [comment]
What do you guys think about this ? I'm interested on coding it ^^

- Luxy.c

**Comment by Dutchman101** [comment]
> What do you guys think about this ? I'm interested on coding it ^^
> 

I gave all of it a read and actually I can't agree less with the complete idea as in how it's written.
I vouch for implementing it without changes to the concept.

Anyone else?

**Comment by qaisjp** [comment]
Everything except number 4 (an event) sounds good to me. I would prefer `getNetworkRequests` and `abortNetworkRequest` as names.

I can see an event being useful for debug purposes but we should consider that in a separate issue as there would also be an event that fires when a request is made.

**Comment by ghost** [comment]
Here are the methods i had planned so far - feedback would be great
```lua
{
    element, -- requestHandle returned from fetch/callRemote
    ...
}
-- returns a list of all pending requests
table getNetworkRequests(string/resource resource)


{
    url      = string, -- host
    name     = string, -- queue name
    resource = resource, -- resource element the request was made from
    start    = number, -- unix timestamp of start
    ...
}
-- returns a table when valid handle was specified, false otherwise
table getRequestInfo(element requestHandle)

-- returns true when request was aborted successfully, false otherwise
bool abortNetworkRequest(element requestHandle)

-- as they were
element/false callRemote(...)
element/false fetchRemote(...)
```
@qaisjp an event called `onNetworkRequestComplete` might be good to dereference handlers (to avoid memory leaks)

- Luxy.c

**Comment by Timic3** [comment]
Anything that would let us abort and track remote requests would be great.

**Comment by MegaThorx** [comment]
What is the current state of this?

It would be great if this could me merged soonish..

**Comment by ghost** [comment]
@MegaThorx, as far as i tested it it's stable enough to be merged. There was only 1 edge case in which the server / client crashed and it was related to filesize (if you exceed around 1,5GB) which is obvious

**Comment by qaisjp** [comment]
Fixed in 8e9d1c6cb2cf5dd8c05966ad2701de5aab1a4d53
